### ISSUE TITLE
Add `severity` secondary option's function support

### ISSUE THREAD

**Issue by xXNickznXx** [stylelint/stylelint#6863]
### What is the problem you're trying to solve?

Title says it all

### What solution would you like to see?

For example:
```json
"at-rule-disallowed-list": {
  "import": {
    "message": "message",
    "severity": "warning"
  },
  "media": {
    "message": "message",
    "severity": "warning"
  }
}
```

**Comment by ybiquitous** [comment]
@xXNickznXx Thanks for the proposal. I understand your motivation.

---

Regarding the `message` option, I believe Stylelint now can satisfy your requirement by using a JS function like this:

```js
module.exports = {
  rules: {
    'at-rule-disallowed-list': [
      ['import', 'media'],
      {
        message: (atRuleName) => {
          return `any message you want with "${atRuleName}"`;
        }
      }
    ]
  }
};
```

You can view the actual behavior with [this demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIAdEAOjOIG4QAaEAMxQBsYA5AQzTkRgA8ONmJMAGdhNcBAB2DAOYIQaaAFdBvLHmH4AvPmCFJ+fNhUxh8XfsOGA5G1wBaY83tQUwto0YQA7jCj3GN1xrcwBtSytDUOt0dWDqfGtOVzZrAF1qCMi9A0irTlE2GRhzAAo7ACUTdk4ASm0APgtcvKtsGFwlbAMAAzZJAE98Avdi-AGIJXxvfoJvFFwACyIQABJgSuqOGABfYh6KLLy9lqsTyLSIk53xSGkUGQAxHDQ7eQArMVpYDDFEHMMxGEuAGzECklwxHMxEYdlMkJoESBILBKAh9jusnswP6UDY2CgUJWsNw8OI+huOyAA).

Although this feature is still experimental, we may be able to publish as non-experimetal since the feature has been provided for a long time.

See also the doc:

https://github.com/stylelint/stylelint/blob/40b4a6daa6168201e55eedae2acef431e216978b/docs/user-guide/configure.md?plain=1#L114-L124

---

Regarding the `severity` option, we cannot configure it like `message`. However, providing a JS function support such as `message` may make sense. For example,

```js
module.exports = {
  rules: {
    'at-rule-disallowed-list': [
      ['import', 'media'],
      {
        severity: (atRuleName) => {
          if (atRuleName === 'import') return 'warning';
          if (atRuleName === 'media') return 'error';
          return null; // respects `defaultSeverity`
        }
      }
    ]
  }
};
```

What do you think?

**Comment by ybiquitous** [comment]
Thanks for the feedback. Firstly, let's support only `severity` to implement it easily. We can add other options' support later if anyone requests.

Requirements:
- The `severity` secondary option can accept a JS function with the same arguments as the `message` secondary option.
- A `severity` function must return one of: `"warning"`, `"error"`, or `null`.
- `null` that a `severity` function returns must mean `defaultSeverity`.
- This feature must be documented as *experimental* for now.

---

I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

**Comment by Mouvedia** [comment]
We currently don't have a distinction between warnings and errors unlike ESLint.
Is the goal to change the exit code returned by the CLI ?

**Comment by Ayc0** [comment]
Would it be possible to be able to return either `null` or `undefined`?
- `undefined` -> apply `defaultSeverity`
- `null` don't throw anything (neither warning nor error)

My use case is that I'd like to use `selector-class-pattern` but with really custom logic, and instead of using regexes (that can be hard to maintain for really complex cases), use a JS function.
And we could use `null` to say "this class is okay, don't report anything"

Maybe we could use `false` instead of `null`. WDYT?

**Comment by ybiquitous** [comment]
@Ayc0 Do you want to disable a rule by `severity`? We can disable a rule with `null` like this:

```js
{
  'selector-class-pattern': null
}
```

https://github.com/stylelint/stylelint/blob/054331783caee2c87bb7491e9aee75a2d9250edb/docs/user-guide/configure.md?plain=1#L30-L32

Either `null` or `undefined` to respect `defaultSeverity` seems OK to me. We may need to accept both `null` and `undefined` for that purpose to reduce people's confusion.

On the other hand, I worry distinguishing `null` and `undefined` with a different meaning may confuse people. 🤔

**Comment by Ayc0** [comment]
> @Ayc0 Do you want to disable a rule by `severity`?

Not really: my intention was to disable the rule based on the value, like for `color-no-hex`, disable it for some values of hex, but keep it enabled for the rest


> On the other hand, I worry distinguishing `null` and `undefined` with a different meaning may confuse people. 🤔

I agree 🤔 
What about an API like this?

```js
module.exports = {
  rules: {
    'at-rule-disallowed-list': [
      ['import', 'media'],
      {
        severity: (atRuleName, defaultSeverity) => {
          if (atRuleName === 'import') return 'warning';
          if (atRuleName === 'media') return 'error';
          if (atRuleName === 'container') return null; // disable rule for @container // This one is a bit dumb, but for rules with regexes, it becomes more powerful
          return defaultSeverity;
        }
      }
    ]
  }
};
```
